internal/iam/repository_grant_scope.go

```diff
@@ -0,0 +1,385 @@
+			}
+			msgs = append(msgs, &roleOplogMsg)
+
+			deleteRoleGrantScopes := make([]any, 0, len(grantScopes))
+			for _, grantScope := range grantScopes {
+				roleGrantScope, err := NewRoleGrantScope(ctx, roleId, grantScope)
+				if err != nil {
+					return errors.Wrap(ctx, err, op, errors.WithMsg("unable to create in memory role grant scope"))
+				}
+				deleteRoleGrantScopes = append(deleteRoleGrantScopes, roleGrantScope)
+			}
+
+			if len(deleteRoleGrantScopes) == 0 {
+				return nil
+			}
+
+			roleGrantScopesOplogMsgs := make([]*oplog.Message, 0, len(deleteRoleGrantScopes))
+			rowsDeleted, err := w.DeleteItems(ctx, deleteRoleGrantScopes, db.NewOplogMsgs(&roleGrantScopesOplogMsgs))
+			if err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("unable to add grants"))
+			}
+			totalRowsDeleted = rowsDeleted
+			msgs = append(msgs, roleGrantScopesOplogMsgs...)
+
+			metadata := oplog.Metadata{
+				"op-type":            []string{oplog.OpType_OP_TYPE_CREATE.String()},
+				"scope-id":           []string{scope.PublicId},
+				"scope-type":         []string{scope.Type},
+				"resource-public-id": []string{role.PublicId},
+			}
+			if err := w.WriteOplogEntryWith(ctx, oplogWrapper, roleTicket, metadata, msgs); err != nil {
+				return errors.Wrap(ctx, err, op, errors.WithMsg("unable to write oplog"))
+			}
+
+			return nil
+		},
+	)
+	if err != nil {
+		return db.NoRowsAffected, errors.Wrap(ctx, err, op)
+	}
+	return totalRowsDeleted, nil
+}
+
+// SetRoleGrantScopes sets grant scopes on a role (roleId). The role's current
+// db version
+// must match the roleVersion or an error will be returned. Zero is not a valid
+// value for the WithVersion option and will return an error.
+func (r *Repository) SetRoleGrantScopes(ctx context.Context, roleId string, roleVersion uint32, grantScopes []string, opt ...Option) ([]*RoleGrantScope, int, error) {
+	const op = "iam.(Repository).SetRoleGrantScopes"
+	if roleId == "" {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing role id")
+	}
+	if roleVersion == 0 {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing version")
+	}
+
+	// Explicitly set to zero clears, but treat nil as a mistake
+	if grantScopes == nil {
+		return nil, db.NoRowsAffected, errors.New(ctx, errors.InvalidParameter, op, "missing grants")
+	}
+
+	reader := r.reader
+	writer := r.writer
+	needFreshReaderWriter := true
+	opts := getOpts(opt...)
+	if opts.withReader != nil && opts.withWriter != nil {
+		reader = opts.withReader
+		writer = opts.withWriter
+		needFreshReaderWriter = false
+	}
+
+	role := allocRole()
+	role.PublicId = roleId
+
+	// NOTE: Set calculation can safely take place out of the transaction since
+	// we are using roleVersion to ensure that we end up operating on the same
+	// set of data from this query to the final set in the transaction function
+
+	// Find existing grant scopes
+	roleGrantScopes := []*RoleGrantScope{}
```

Comment:
Like AddRoleGrantScopes, this read and subsequent work needs to move into the Tx after you've gotten the Ticket for the role.

Reply:
We use the passed-in `roleVersion` so it should be fine.